### PR TITLE
Add standardization/write to BQ step for CTP

### PR DESCRIPTION
**Pull request by jenniebrown**
- Implements write_to_bq for the CTP data source
- Adds two standardized race categories, API and INDIGENOUS
- Adds unit tests

#116

### CODE REVIEW

File: python/datasources/covid_tracking_project.py
Comment (mayaspivak) [inline python/datasources/covid_tracking_project.py:21]: Add method comment here and to other helper methods below.

**Review by mayaspivak — APPROVED** [review]

File: python/datasources/covid_tracking_project.py
Comment (aaronsn) [inline python/datasources/covid_tracking_project.py:59]: Per a comment in some google doc thread, don't we need a `reports_ethnicity` column too? I'm forgetting the exact details, but I think we discussed needing that information in order to know how to join with ACS?

File: python/datasources/covid_tracking_project.py
Comment (aaronsn) [inline python/datasources/covid_tracking_project.py:78]: Some general data cleanup stuff:
- Could we convert the dates to `yyyy-mm-dd` format?
- Not sure if CTP fixed this issue, but in the past when I was running a script on their data, some of the int values were actually strings with commas in them, so I had to [fix those](https://github.com/aaronsn/health-equity-tracker/blob/566860cf99fe1fd3be65ca930db1c677d9efcdbf/python/datasources/covid_tracking_project.py#L103) before uploading to BQ

File: python/datasources/covid_tracking_project.py
Comment (aaronsn) [inline python/datasources/covid_tracking_project.py:31]: I think maybe we should split OTHER into two (not sure on the exact naming):
1. The same "Other" that matches up with ACS "Other"
2. A different "Other" that can't be joined with ACS

If a dataset uses the exact same race categories as ACS then we can join the "other" group, otherwise we can't. Even if we mostly just use the non-joinable one, I think being explicit would be better

File: python/datasources/covid_tracking_project.py
Comment (aaronsn) [inline python/datasources/covid_tracking_project.py:55]: Use the constant from the standardized columns file?

File: python/datasources/covid_tracking_project.py
Comment (aaronsn) [inline python/datasources/covid_tracking_project.py:21]: Are you still planning on splitting `race_and_ethnicity` into two columns (`race` and `includes_hispanic`)? I wasn't sure if that was intended as a followup or part of this

**Review by aaronsn — CHANGES_REQUESTED** [review]
I really like how you're doing the metadata joining. We should be able to reuse this pattern to get fips codes into tables during the standardization step in the future

File: python/datasources/covid_tracking_project.py
Comment (jenniebrown) [reply python/datasources/covid_tracking_project.py:72]: There's only one state that does this right now, and they do report under AIAN. If other states start to do this (I think it's unlikely) then we'll have to double check.

File: python/ingestion/gcs_to_bq_util.py
Comment (jenniebrown) [reply python/ingestion/gcs_to_bq_util.py:179]: The warning about windows is just for dev environments, but the first comment flags a potential race condition/memory concern for all environments. I added this comment a while ago since the call to os.remove() has behavior that could be unexpected (I originally thought that the memory referenced by the resulting dataframe could be corrupted when the file was removed, but that is not the case in Unix). I'm not sure when this comment was removed but it seems like something that is being added back in due to a desync in my workspace. I'd like to keep it in if that's all right. 

I personally think that this is kind of a code smell and would prefer to fix it, but I think it might require a bit of a refactor.

File: python/datasources/covid_tracking_project.py
Comment (jenniebrown) [reply python/datasources/covid_tracking_project.py:52]: :)

File: python/datasources/covid_tracking_project.py
Comment (jenniebrown) [reply python/datasources/covid_tracking_project.py:59]: Yes, that information is included in the metadata table.

File: python/datasources/covid_tracking_project.py
Comment (jenniebrown) [reply python/datasources/covid_tracking_project.py:21]: Yes, when the metadata is joined the includes_hispanic column is added. I used the column name race_and_ethnicity still because it is a standard across multiple data sets, and it will still include ethnicity alone as its own "race category" (Hispanic and Non Hispanic), if I understood correctly.

File: python/datasources/covid_tracking_project.py
Comment (jenniebrown) [reply python/datasources/covid_tracking_project.py:65]: No reason, they are equivalent

File: python/datasources/covid_tracking_project.py
Comment (jenniebrown) [reply python/datasources/covid_tracking_project.py:78]: Yeah, I'll format the dates.

I haven't been able to repro the ints as strings issue you describe. The CTP data is a CSV, so the data is untyped and pandas has been importing them as ints

File: python/datasources/covid_tracking_project.py
Comment (jenniebrown) [reply python/datasources/covid_tracking_project.py:75]: Yeah, this is the big one that I wanted to get your input on. Technically any of the metadata columns can vary between cases/deaths/hospitalizations/tests, and several of them do (e.g. race_ethnicity_combined, reports_ethnicity, reports_race, uses_api...) This will result in a proliferation of columns if we use a pivoted table. My reasoning for using cases/deaths/etc as a breakdown column instead was that it makes it easy to look up the data you want without needing to do string parsing on column names. It also makes joining with the ACS data more straightforward.

I haven't worked on the frontend at all, but based on Krista's talk it seems like we would need to add a new Breakdown class? I'm not sure how much work that would be but in my head this format seems easier to work with. If it's not feasible, I think that creating multiple tables might be slightly preferable

File: python/datasources/covid_tracking_project.py
Comment (jenniebrown) [reply python/datasources/covid_tracking_project.py:79]: Oooh okay I figured out the strings issue. I found if you set thousands=',' in the call to pd.read_csv, it will parse the numbers correctly. Woo!!

File: python/ingestion/gcs_to_bq_util.py
Comment (aaronsn) [reply python/ingestion/gcs_to_bq_util.py:183]: Oh yeah definitely fine to keep the comment, I just meant it would be nice to add something like "This shouldn't affect production environments but may be an issue for local dev on Windows"

File: python/datasources/covid_tracking_project.py
Comment (aaronsn) [reply python/datasources/covid_tracking_project.py:76]: I agree with the part about not having separate columns because joining with ACS gets complicated and I don't want to do string-parsing on column names.

A 4th alternative I just thought of is using nulls and add additional rows where they differ:
```
reports_combined | cases | deaths
true             | 400   | null
false            | null  | 20
```
This would allow joining with ACS smoothly, and we just filter out nulls first. Thoughts on that?

@kkatzen what do you think? To clarify, the format Jennie is suggesting here is:
```
state | race_and_ethnicity | variable | number
NY    | white              | cases    | 50
NY    | white              | deaths   | 50
NY    | white              | hosp     | 50
```
This would require some massaging on the frontend. I don't think it'd be terribly complicated to do that in the variable providers. But I'm hesitant to add more custom logic on the frontend than already exists.

My preference is this 4th alternative using additional rows, and my second preference is separate files. What do y'all think?

File: python/datasources/covid_tracking_project.py
Comment (aaronsn) [reply python/datasources/covid_tracking_project.py:79]: wooooah. That's pretty neat, nice find! Of course pandas can do this automagically

File: python/datasources/covid_tracking_project.py
Comment (aaronsn) [reply python/datasources/covid_tracking_project.py]: Sure, fine to not block on this. If there's a possibility you don't get to this before the end of the fellowship, do you mind filing a github issue in case you don't finish?

I think the "defines_other" doesn't necessarily give all the information we need to know? Even if "other" is defined, it still needs to be defined in the exact same way as the ACS "other" category, which may not always be the case?

File: python/datasources/covid_tracking_project.py
Comment (aaronsn) [reply python/datasources/covid_tracking_project.py:22]: Oh, I thought the way you were suggesting was that Hispanic/Non-hispanic would be part of the "race" column? I may have forgotten the details here...

**Review by aaronsn — APPROVED** [review]
Approving to unblock you - up to you how to chunk the followups we discussed as different PRs

File: python/datasources/covid_tracking_project.py
Comment (jenniebrown) [reply python/datasources/covid_tracking_project.py:76]: As discussed offline, we will go for option 2, splitting into multiple tables. I opened issue #308

File: python/datasources/covid_tracking_project.py
Comment (jenniebrown) [reply python/datasources/covid_tracking_project.py]: Created #309

File: python/datasources/covid_tracking_project.py
Comment (jenniebrown) [reply python/datasources/covid_tracking_project.py:22]: Renamed to "race"

**Comment by jenniebrown** [comment]
Thanks for the reviews! I created issues for the items we discussed

**Pull request merged**
